File: java/longest-substring-without-repeating-char.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        if (s.length() <= 1)
            return s.length();
        int i = 0, j = 0, ans = Integer.MIN_VALUE;
        HashMap<Character, Integer> mp = new HashMap<Character, Integer>();
        while (j < s.length()) {
            mp.merge(s.charAt(j), 1, Integer::sum);
            if (mp.size() == j - i + 1) {
                ans = Math.max(ans, j - i + 1);
            }
            else {
                while (mp.size() < j - i + 1) {
                    mp.merge(s.charAt(i), -1, Integer::sum);
                    int v = mp.get(s.charAt(i));
                    if (v == 0)
                        mp.remove(s.charAt(i));
                    i++;
                }
            }
            j++;
        }
        return ans;
    }
}

File: cpp/longest-substring-without-repeating-char.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        if (s.size() <= 1)
            return s.size();
        int i = 0, j = 0, ans = INT_MIN;
        unordered_map<char, int> mp;
        while (j < s.size()) {
            mp[s[j]]++;
            
            if (mp.size() == j - i + 1) {
                ans = max(ans, j - i + 1);
                j++;
            }
            else {
                while (mp.size() < j - i + 1) {
                    mp[s[i]]--;
                    if (mp[s[i]] == 0)
                        mp.erase(s[i]);
                    i++;
                }
                j++;
            }
        }
        return ans;
    }
};
